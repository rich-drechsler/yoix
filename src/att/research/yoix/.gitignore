#
# Ignore class and jar files, unless told to do otherwise.
#

*.class
*.jar

#
# Unfortunately there are a few class files that are (or were) part of the Yoix
# package, so we don't ignore them.
#

!apple/STORAGE/**
!jnlp/STORAGE/**

#
# These automatically generated files should always be ignored.
#

CharStream.java
DOTParser.java
DOTParserConstants.java
DOTParserTokenManager.java
DOTParserTreeConstants.java
JJTDOTParserState.java
JJTPatternParserState.java
JJTXMLParserState.java
JJTYoixParserState.java
Node.java
ParseException.java
PatternParser.java
PatternParserConstants.java
PatternParserTokenManager.java
PatternParserTreeConstants.java
Token.java
TokenMgrError.java
XMLParser.java
XMLParserConstants.java
XMLParserTokenManager.java
XMLParserTreeConstants.java
YoixParser.java
YoixParserConstants.java
YoixParserTokenManager.java
YoixParserTreeConstants.java
