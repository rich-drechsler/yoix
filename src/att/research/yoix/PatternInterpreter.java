/*
 *  This software may only be used by you under license from AT&T Corp.
 *  ("AT&T").  A copy of AT&T's Source Code Agreement is available at
 *  AT&T's Internet website having the URL:
 *
 *    <http://www.research.att.com/sw/tools/yoix/license/source.html>
 *
 *  If you received this software without first entering into a license
 *  with AT&T, you have an infringing copy of this software and cannot
 *  use it without violating AT&T's intellectual property rights.
 */

package att.research.yoix;
import java.io.*;

abstract
class PatternInterpreter

    implements YoixConstants

{

    //
    // A first cut at an interpreter for pattern matching operations that
    // are described in a parse tree that was built by PatternParser. Much
    // of this was borrowed from the yoix interpreter, so there's probably
    // room for simplification - later.
    //
    // NOTE - names of Yoix and pattern parser constants collide but their
    // values don't, so be careful!! That's also why we couldn't implement
    // both YoixParserConstants and PatternParserConstants.
    //
    // NOTE - the actual matching is handled in matchValues(), so that may
    // be where to start looking if you decide to change things.
    //

    ///////////////////////////////////
    //
    // PatternInterpreter Methods
    //
    ///////////////////////////////////

    static boolean
    match(SimpleNode pattern, String arg, int flags) { 

	YoixStack  stack = VM.getThreadStack();

	matchPattern(pattern, arg, flags|SINGLE_BYTE, stack);
	return(stack.popYoixObject().booleanValue());
    }

    ///////////////////////////////////
    //
    // Private Methods
    //
    ///////////////////////////////////

    private static int
    conditional(SimpleNode node, int n, String arg, int flags, YoixStack stack) {

	SimpleNode  cond;
	YoixObject  expr;

	cond = node.getChild(n);

	if (cond.type() == PatternParserConstants.CONDITIONAL) {
	    expr = stack.popRvalue();
	    switch (node.getChild(++n).type()) {
		case PatternParserConstants.AND:
		    stack.pushInt(expr.booleanValue() && match(cond.getChild0(), arg, flags));
		    break;

		case PatternParserConstants.OR:
		    stack.pushInt(expr.booleanValue() || match(cond.getChild0(), arg, flags));
		    break;

		default:
		    VM.abort(INTERNALERROR);
		    break;
	    }
	} else VM.abort(INTERNALERROR);

	return(n);
    }


    private static void
    matchPattern(SimpleNode node, String arg, int flags, YoixStack stack) {

	SimpleNode  child;
	int         length;
	int         op;
	int         n;

	if (arg != null && node != null) {
	    if (node.type() == PatternParserConstants.PATTERN) {
		if ((length = node.length()) > 0) {
		    for (n = 0; n < length; n++) {
			child = node.getChild(n);
			switch (op = child.type()) {
			    case PatternParserConstants.CONDITIONAL:
				n = conditional(node, n, arg, flags, stack);
				break;

			    case PatternParserConstants.NOT:
				stack.pushInt(!stack.popRvalue().booleanValue());
				break;

			    case PatternParserConstants.PATTERN:
				matchPattern(child, arg, flags, stack);
				break;

			    case PatternParserConstants.XOR:
				stack.pushInt(stack.popRvalue().booleanValue() ^ stack.popRvalue().booleanValue());
				break;

			    default:
				if (child instanceof YoixObject)
				    stack.pushInt(matchValues((YoixObject)child, arg, flags, stack));
				else VM.abort(INTERNALERROR);
				break;
			}
		    }
		} else stack.pushInt(false);
	    } else stack.pushInt(false);
	} else stack.pushInt(false);
    }


    private static boolean
    matchValues(YoixObject obj, String arg, int flags, YoixStack stack) {

	YoixRERegexp  regexp;
	boolean       result;
	String        pattern;
	byte          source[];
	byte          values[];
	int           length;
	int           n;

	//
	// Some of this, particularly the matching triggered when obj is
	// a string and DOT_PATTERN is set, may be convenient, but it was
	// added primarily to support an existing internal project. You
	// should be able to use regular expressions to accomplish the
	// same thing.
	//

	if (obj != null && arg != null) {
	    if (obj.isString()) {
		pattern = obj.stringValue();
		if ((flags & DOT_PATTERN) != 0) {
		    if ((length = arg.length()) == pattern.length()) {
			result = true;
			if ((flags & CASE_INSENSITIVE) != 0) {
			    arg = arg.toLowerCase();
			    pattern = pattern.toLowerCase();
			}
			source = YoixMake.javaByteArray(arg);
			values = YoixMake.javaByteArray(pattern);
			for (n = 0; n < length; n++) {
			    if (source[n] != values[n] && values[n] != '.') {
				result = false;
				break;
			    }
			}
		    } else result = false;
		} else if ((flags & CASE_INSENSITIVE) != 0)
		    result = arg.equalsIgnoreCase(pattern);
		else result = arg.equals(pattern);
	    } else if (obj.isRegexp()) {
		//
		// Important to check and set the type first. Also note
		// that flags currently always has SINGLE_BYTE set.
		//
		if (obj.getInt(N_TYPE, SINGLE_BYTE) != (flags&REFLAGS_MASK))
		    obj.putInt(N_TYPE, flags&REFLAGS_MASK);
		if ((regexp = (YoixRERegexp)obj.getManagedObject()) != null)
		    result = regexp.exec(arg, null);
		else result = false;
	    } else result = false;
	} else result = false;

	return(result);
    }
}

