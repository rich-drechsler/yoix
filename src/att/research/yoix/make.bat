@echo off

rem
rem This is a batch file that may help if you're stuck on a Windows PC
rem with our source package. You need to install Sun's Java Development
rem Kit for Windows and WebGain's JavaCC Parser Generator before this
rem batch file will run. JDK is available at www.javasoft.com (get the
rem latest 1.X release), while JavaCC (Version 1.2) can be downloaded
rem from www.webgain.com. Both packages are free and easy to install.
rem Once they're on your system update our PATH definition (right after
rem the comments) and that should be it.
rem
rem Typing,
rem
rem	make all
rem
rem builds class files from our source package, while
rem
rem	make clean
rem
rem removes class files and any intermediate files created by jjtree and
rem javacc. Type
rem
rem	make yoix.jar
rem
rem to build a jar file.
rem 

set OPATH=%PATH%
set PATH=\JavaCC\bin;\JavaSoft\JDK117B\bin;%PATH%

set CLASSPATH=..\..\..

if "%1"=="" goto ALL
if "%1"=="all" goto ALL
if "%1"=="clean" goto CLEAN
if "%1"=="clobber" goto CLOBBER
if "%1"=="yoix.jar" goto JAR
goto DONE

:ALL
	echo 	jjtree XMLParser.jjt
	call jjtree XMLParser.jjt

	echo 	javacc XMLParser.jj
	call javacc XMLParser.jj

	echo 	jjtree YoixParser.jjt
	call jjtree YoixParser.jjt

	echo 	javacc YoixParser.jj
	call javacc YoixParser.jj

	echo 	javac XMLParser.java
	call javac XMLParser.java

	echo 	javac YoixMain.java
	call javac YoixMain.java

	echo 	javac YoixMiscCtype.java
	call javac YoixMiscCtype.java

	echo 	javac YoixMiscPrintf.java
	call javac YoixMiscPrintf.java

	echo 	javac YoixMiscScanf.java
	call javac YoixMiscScanf.java

	echo 	javac YoixModuleAWT.java
	call javac YoixModuleAWT.java

	echo 	javac YoixModuleCalendar.java
	call javac YoixModuleCalendar.java

	echo 	javac YoixModuleCtype.java
	call javac YoixModuleCtype.java

	echo 	javac YoixModuleDate.java
	call javac YoixModuleDate.java

	echo 	javac YoixModuleError.java
	call javac YoixModuleError.java

	echo 	javac YoixModuleGraph.java
	call javac YoixModuleGraph.java

	echo 	javac YoixModuleGraphics.java
	call javac YoixModuleGraphics.java

	echo 	javac YoixModuleImage.java
	call javac YoixModuleImage.java

	echo 	javac YoixModuleIO.java
	call javac YoixModuleIO.java

	echo 	javac YoixModuleLocale.java
	call javac YoixModuleLocale.java

	echo 	javac YoixModuleMath.java
	call javac YoixModuleMath.java

	echo 	javac YoixModuleMisc.java
	call javac YoixModuleMisc.java

	echo 	javac YoixModuleNet.java
	call javac YoixModuleNet.java

	echo 	javac YoixModuleParser.java
	call javac YoixModuleParser.java

	echo 	javac YoixModuleRE.java
	call javac YoixModuleRE.java

	echo 	javac YoixModuleSound.java
	call javac YoixModuleSound.java

	echo 	javac YoixModuleStdio.java
	call javac YoixModuleStdio.java

	echo 	javac YoixModuleStream.java
	call javac YoixModuleStream.java

	echo 	javac YoixModuleString.java
	call javac YoixModuleString.java

	echo 	javac YoixModuleSystem.java
	call javac YoixModuleSystem.java

	echo 	javac YoixModuleThread.java
	call javac YoixModuleThread.java

	echo 	javac YoixModuleType.java
	call javac YoixModuleType.java

	echo 	javac YoixModuleUtil.java
	call javac YoixModuleUtil.java

	echo 	javac YoixModuleWindows.java
	call javac YoixModuleWindows.java

	echo 	javac YoixPublic.java
	call javac YoixPublic.java

	echo 	javac YoixPublicConstants.java
	call javac YoixPublicConstants.java

	goto DONE

:CLEAN
	echo dummy file >dummy.class

	echo 	del *.class
	del *.class

	echo 	del JJTYoixParserState.java
	if exist JJTYoixParserState.java del JJTYoixParserState.java

	echo 	del YoixParser.jj
	if exist YoixParser.jj del YoixParser.jj

	echo 	del YoixParser.java
	if exist YoixParser.java del YoixParser.java

	echo 	del YoixParserConstants.java
	if exist YoixParserConstants.java del YoixParserConstants.java

	echo 	del YoixParserTokenManager.java
	if exist YoixParserTokenManager.java del YoixParserTokenManager.java

	echo 	del YoixParserTreeConstants.java
	if exist YoixParserTreeConstants.java del YoixParserTreeConstants.java

	echo 	del JJTXMLParserState.java
	if exist JJTXMLParserState.java del JJTXMLParserState.java

	echo 	del XMLParser.jj
	if exist XMLParser.jj del XMLParser.jj

	echo 	del XMLParser.java
	if exist XMLParser.java del XMLParser.java

	echo 	del XMLParserConstants.java
	if exist XMLParserConstants.java del XMLParserConstants.java

	echo 	del XMLParserTokenManager.java
	if exist XMLParserTokenManager.java del XMLParserTokenManager.java

	echo 	del XMLParserTreeConstants.java
	if exist XMLParserTreeConstants.java del XMLParserTreeConstants.java

	echo 	del Node.java
	if exist Node.java del Node.java

	echo 	del ParseException.java
	if exist ParseException.java del ParseException.java

	echo 	del Token.java
	if exist Token.java del Token.java

	echo 	del TokenMgrError.java
	if exist TokenMgrError.java del TokenMgrError.java

	goto DONE

:CLOBBER
	echo 	del yoix.jar
	if exist yoix.jar del yoix.jar

	goto CLEAN

:JAR
	if not exist YoixMain.class call %0

	cd ..\..\..
	echo 	jar cvf att/research/yoix/yoix.jar att/research/yoix/*.class
	jar cvf att/research/yoix/yoix.jar att/research/yoix/*.class
	cd att\research\yoix

	goto DONE

:DONE
	set PATH=%OPATH%
