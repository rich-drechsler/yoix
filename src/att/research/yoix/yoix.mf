Main-Class: att.research.yoix.YoixMain
