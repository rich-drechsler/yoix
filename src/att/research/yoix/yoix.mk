#
# Yoix makefile
#

ROOT = ../../..
MAKEFILE = $(firstword $(MAKEFILE_LIST))

include $(ROOT)/make/common.mk

CLASSPATH = $(ROOT)

#
# YOIX_PACKAGES names the additional packages that will be included in the Yoix
# jar file. The list doesn't include everything that's available in the source
# (e.g., j3d).
#

YOIX_PACKAGES = ydat ychart jvma apple

#
# The Yoix parser should be last.
#

PARSERS = \
	DOTParser.jjt \
	PatternParser.jjt \
	XMLParser.jjt \
	YoixParser.jjt

SOURCE = \
	Yoix.java \
	YoixAPI.java \
	YoixAPIProtected.java \
	YoixAWTCanvas.java \
	YoixAWTCheckbox.java \
	YoixAWTCheckboxMenuItem.java \
	YoixAWTCheckboxMenuItemGroup.java \
	YoixAWTDialog.java \
	YoixAWTErrorDialog.java \
	YoixAWTFileDialog.java \
	YoixAWTFontMetrics.java \
	YoixAWTFrame.java \
	YoixAWTGridLayout.java \
	YoixAWTInvocationEvent.java \
	YoixAWTList.java \
	YoixAWTOwnerFrame.java \
	YoixAWTPanel.java \
	YoixAWTTableColumn.java \
	YoixAWTTableManager.java \
	YoixAWTTextArea.java \
	YoixAWTTextCanvas.java \
	YoixAWTTextComponent.java \
	YoixAWTTextField.java \
	YoixAWTTextTerm.java \
	YoixAWTTexturePaint.java \
	YoixAWTToolkit.java \
	YoixAWTWindow.java \
	YoixBinder.java \
	YoixBodyArray.java \
	YoixBodyAudioClip.java \
	YoixBodyBlock.java \
	YoixBodyBuiltin.java \
	YoixBodyCalendar.java \
	YoixBodyCertificate.java \
	YoixBodyCipher.java \
	YoixBodyClipboard.java \
	YoixBodyCompiler.java \
	YoixBodyComponent.java \
	YoixBodyComponentAWT.java \
	YoixBodyComponentSwing.java \
	YoixBodyCookieManager.java \
	YoixBodyControl.java \
	YoixBodyDatagramSocket.java \
	YoixBodyDictionary.java \
	YoixBodyDictionaryObject.java \
	YoixBodyDictionaryThis.java \
	YoixBodyElement.java \
	YoixBodyFont.java \
	YoixBodyFunction.java \
	YoixBodyGraphObserver.java \
	YoixBodyGraphics.java \
	YoixBodyHashtable.java \
	YoixBodyImage.java \
	YoixBodyJump.java \
	YoixBodyKey.java \
	YoixBodyKeyStore.java \
	YoixBodyLocale.java \
	YoixBodyMatrix.java \
	YoixBodyMulticastSocket.java \
	YoixBodyNull.java \
	YoixBodyNumber.java \
	YoixBodyOption.java \
	YoixBodyParseTree.java \
	YoixBodyPath.java \
	YoixBodyProcess.java \
	YoixBodyRandom.java \
	YoixBodyRegexp.java \
	YoixBodyScreen.java \
	YoixBodySecurityManager.java \
	YoixBodyServerSocket.java \
	YoixBodySocket.java \
	YoixBodyStream.java \
	YoixBodyString.java \
	YoixBodySubexp.java \
	YoixBodyTag.java \
	YoixBodyThread.java \
	YoixBodyTimeZone.java \
	YoixBodyTransferHandler.java \
	YoixBodyUIManager.java \
	YoixBodyVector.java \
	YoixBodyZipEntry.java \
	YoixChecksum.java \
	YoixClassLoader.java \
	YoixCoderInputStream.java \
	YoixCoderOutputStream.java \
	YoixCompiler.java \
	YoixCompilerConstants.java \
	YoixCompilerSupport.java \
	YoixCompilerSyncLocals.java \
	YoixConstants.java \
	YoixConstantsAWT.java \
	YoixConstantsError.java \
	YoixConstantsErrorName.java \
	YoixConstantsGraph.java \
	YoixConstantsGraphics.java \
	YoixConstantsImage.java \
	YoixConstantsJFC.java \
	YoixConstantsJTree.java \
	YoixConstantsJTable.java \
	YoixConstantsStream.java \
	YoixConstantsSwing.java \
	YoixConstantsXColor.java \
	YoixConverter.java \
	YoixConverterInput.java \
	YoixConverterOutput.java \
	YoixDataInputStream.java \
	YoixDataOutputStream.java \
	YoixDataTransfer.java \
	YoixDragManager.java \
	YoixError.java \
	YoixFontType0.java \
	YoixGraphBase.java \
	YoixGraphElement.java \
	YoixImageObserver.java \
	YoixInterfaceBody.java \
	YoixInterfaceCallable.java \
	YoixInterfaceCloneable.java \
	YoixInterfaceDnD.java \
	YoixInterfaceDragable.java \
	YoixInterfaceDrawable.java \
	YoixInterfaceFileChooser.java \
	YoixInterfaceFont.java \
	YoixInterfaceKillable.java \
	YoixInterfaceListener.java \
	YoixInterfaceMenuBar.java \
	YoixInterfacePointer.java \
	YoixInterfaceShowing.java \
	YoixInterfaceSortable.java \
	YoixInterfaceWindow.java \
	YoixInterpreter.java \
	YoixInterruptable.java \
	YoixMake.java \
	YoixMakeEvent.java \
	YoixMakeScreen.java \
	YoixMain.java \
	YoixMisc.java \
	YoixMiscCtype.java \
	YoixMiscGeom.java \
	YoixMiscGraph.java \
	YoixMiscGraphics.java \
	YoixMiscJFC.java \
	YoixMiscMenu.java \
	YoixMiscPrintf.java \
	YoixMiscQsort.java \
	YoixMiscScanf.java \
	YoixMiscSSL.java \
	YoixMiscTime.java \
	YoixMiscXML.java \
	YoixModule.java \
	YoixModuleAWT.java \
	YoixModuleBorder.java \
	YoixModuleCalendar.java \
	YoixModuleCtype.java \
	YoixModuleDate.java \
	YoixModuleError.java \
	YoixModuleEvent.java \
	YoixModuleFactorial.java \
	YoixModuleGraph.java \
	YoixModuleGraphics.java \
	YoixModuleImage.java \
	YoixModuleIO.java \
	YoixModuleJFC.java \
	YoixModuleLayout.java \
	YoixModuleLocale.java \
	YoixModuleMath.java \
	YoixModuleMisc.java \
	YoixModuleNet.java \
	YoixModuleParser.java \
	YoixModuleRE.java \
	YoixModuleRobot.java \
	YoixModuleSecure.java \
	YoixModuleSound.java \
	YoixModuleStdio.java \
	YoixModuleStream.java \
	YoixModuleString.java \
	YoixModuleSwing.java \
	YoixModuleSwingExtension.java \
	YoixModuleSystem.java \
	YoixModuleThread.java \
	YoixModuleType.java \
	YoixModuleUtil.java \
	YoixModuleVM.java \
	YoixModuleWindows.java \
	YoixModuleXColor.java \
	YoixObject.java \
	YoixOption.java \
	YoixParserBvalue.java \
	YoixParserStream.java \
	YoixPointer.java \
	YoixPointerActive.java \
	YoixReflect.java \
	YoixRERegexp.java \
	YoixRESubexp.java \
	YoixRegistryColor.java \
	YoixRegistryCursor.java \
	YoixSecurityManager.java \
	YoixSecurityOptions.java \
	YoixSimpleDateFormat.java \
	YoixSplashScreen.java \
	YoixStack.java \
	YoixSwingDefaultCaret.java \
	YoixSwingHighlighter.java \
	YoixSwingJCanvas.java \
	YoixSwingJColorChooser.java \
	YoixSwingJComboBox.java \
	YoixSwingJDesktopPane.java \
	YoixSwingJDialog.java \
	YoixSwingJFileChooser.java \
	YoixSwingJFileDialog.java \
	YoixSwingJFrame.java \
	YoixSwingJInternalFrame.java \
	YoixSwingJLayeredPane.java \
	YoixSwingJList.java \
	YoixSwingJPanel.java \
	YoixSwingJScrollBar.java \
	YoixSwingJScrollPane.java \
	YoixSwingJSlider.java \
	YoixSwingJSplitPane.java \
	YoixSwingJTabbedPane.java \
	YoixSwingJTable.java \
	YoixSwingJTextArea.java \
	YoixSwingJTextCanvas.java \
	YoixSwingJTextComponent.java \
	YoixSwingJTextField.java \
	YoixSwingJTextPane.java \
	YoixSwingJTextTerm.java \
	YoixSwingJTree.java \
	YoixSwingJWindow.java \
	YoixSwingLabelItem.java \
	YoixSwingTableColumn.java \
	YoixSwingTransferHandler.java \
	YoixThread.java \
	YoixTipManager.java \
	YoixTrustPolicy.java \
	YoixUtilHashtable.java \
	YoixUtilVector.java \
	YoixVM.java \
	YoixVMCleaner.java \
	YoixVMClipboard.java \
	YoixVMDisposer.java \
	YoixVMError.java \
	YoixVMShutdownThread.java \
	YoixVMThread.java \
	YoixVMThreadData.java \
	SimpleNode.java \
	PatternInterpreter.java

SUPPORT = \
	ExampleModule.java \
	YoixInstaller.java \
	yoix_linux.bash \
	yoix.bat

all clean clobber install :
	@#
	@# Made lots of changes to this file. Not certain if we need to explicitly
	@# export variables - seems to work as is, but probably needs testing.
	@#
	@echo "++++ Making $@ for the yoix interpreter ++++"
	@$(MAKE) -e -f $(MAKEFILE) local_$@
	@for PACKAGE in $(YOIX_PACKAGES); do \
	    echo "++++ Making $@ for the $$PACKAGE package ++++"; \
	    (cd $$PACKAGE && $(MAKE) -e -f $${PACKAGE}.mk $@) || exit 1; \
	done

run : all
	@#
	@# During development we often type something like
	@#
	@#     make -f yoix.mk RUNARGS="Tests/components3.yx" run
	@#
	@# to run the interpreter using the source code in this directory. Bottom
	@# line is everything assigned to RUNARGS is passed to the interpreter.
	@#
	CLASSPATH=$(CLASSPATH) $(JAVA) $(JAVAFLAGS) att.research.yoix.YoixMain $(RUNARGS)

#
# Local (i.e., for this directory) versions of some of the standard targets.
#

local_all : $(PARSERS:.jjt=.java) $(PARSERS:.jjt=.class) $(SOURCE:.java=.class)
	@:

local_clean :
	rm -f *.class

local_clobber : local_clean
	rm -f *.jar
	@#
	@# Quietly remove files that were automatically generated by javacc and
	@# jjtree. This list may need to be periodically updated.
	@#
	@rm -f Node.java ParseException.java Token.java TokenMgrError.java
	@rm -f YoixParser.java YoixParser*Constants.java YoixParserTokenManager.java
	@rm -f JJTYoixParser*.java YoixParser*.jj
	@rm -f *DOTParser*.java DOTParser*.jj
	@rm -f *XMLParser*.java XMLParser*.jj
	@rm -f *PatternParser*.java PatternParser*.jj
	@rm -f ASCII_CharStream.java ASCII_UCodeESC_CharStream.java CharStream.java

local_install : yoix.jar yoix_linux.bash yoix.bat
	@#
	@# Recent addition that can be used to install yoix.jar and scripts that
	@# run the interpreter on several operating systems (e.g., Linux, Windows).
	@# Nothing is installed if INSTALLDIR is empty or doesn't exist.
	@#
	@if [ "$(INSTALLDIR)" ]; \
	    then \
		if [ -d "$(INSTALLDIR)" ]; \
		    then \
			echo "install -d -m755 $(INSTALLDIR)/lib"; \
			install -d -m755 $(INSTALLDIR)/lib; \
			echo "install -m644 yoix.jar $(INSTALLDIR)/lib"; \
			install -m644 yoix.jar $(INSTALLDIR)/lib; \
			echo "install -d -m755 $(INSTALLDIR)/bin-linux"; \
			install -d -m755 $(INSTALLDIR)/bin-linux; \
			echo "install -m755 yoix_linux.bash $(INSTALLDIR)/bin-linux/yoix"; \
			install -m755 yoix_linux.bash $(INSTALLDIR)/bin-linux/yoix; \
			echo "install -d -m755 $(INSTALLDIR)/bin-windows"; \
			install -d -m755 $(INSTALLDIR)/bin-windows; \
			echo "install -m644 yoix.bat $(INSTALLDIR)/bin-windows/yoix.bat"; \
			install -m644 yoix.bat $(INSTALLDIR)/bin-windows/yoix.bat; \
		    else echo "Skipping install because installation directory $(INSTALLDIR) doesn't exist"; \
		fi; \
	    else \
		if [ "$(INSTALLDIR_ORIGINAL)" ]; \
		    then echo "Skipping install because installation directory $(INSTALLDIR_ORIGINAL) doesn't exist"; \
		    else echo "Skipping install because no installation directory is set"; \
		fi; \
	fi

#
# Class files from all the supported packages end up in the yoix.jar file. It's
# just not worth the effort to track them all here, so we just rebuild yoix.jar
# whenever this rule is triggered. Ugly, but good enough for now.
#

yoix.jar : all
	@rm -f $@
	cd $(ROOT); $(JAR) cvfm att/research/yoix/$@ att/research/yoix/yoix.mf att/research/yoix/*.class
	@for DIR in $(YOIX_PACKAGES); do \
	    (cd $(ROOT) && $(JAR) vfu att/research/yoix/$@ att/research/yoix/$$DIR/*.class) || exit 1; \
	done
	cd $(ROOT); $(JAR) vfu att/research/yoix/$@ att/research/yoix/resources/system/*.yx
	$(JAR) i $@

