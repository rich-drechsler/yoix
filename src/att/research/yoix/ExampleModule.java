/*
 *  This software may only be used by you under license from AT&T Corp.
 *  ("AT&T").  A copy of AT&T's Source Code Agreement is available at
 *  AT&T's Internet website having the URL:
 *
 *    <http://www.research.att.com/sw/tools/yoix/license/source.html>
 *
 *  If you received this software without first entering into a license
 *  with AT&T, you have an infringing copy of this software and cannot
 *  use it without violating AT&T's intellectual property rights.
 */

import att.research.yoix.*;	// see the NOTE below

//
// An example of a custom module that defines a few trivial builtins
// and is automatically loaded when it's referenced in standard dot
// notation that begins with "yoix.module." and ends with the full
// name of this class. In addition this class file must be put in
// directory that Java examines when it searches for a class using
// the CLASSPATH variable.
//
// For example, if you're in the Yoix source directory type,
//
//	make all ExampleModule.class
//
// to build the interpreter and ExampleModule.class, then type
//
//	CLASSPATH=../../..:. java  att.research.yoix.YoixMain -d1
//
// to run yoix (notice the :. at the end of CLASSPATH), and finally
// type
//
//	yoix.module.ExampleModule.version();
//
// and the interpreter should automatically load ExampleModule, run
// the version() builtin that's defined below, and print the answer
// on standard output (because of the -d1 flag).
//
// Although this works, user module source code is not usually kept
// with the Yoix source code. We've written at least one substantial
// user module, and in every case we kept the module source code in
// a completely separate directory.
//
// Custom module support first appeared in version 0.9.6 even though
// it all worked properly at that time there were some details that
// needed attention - we plan on tackling the missing pieces soon!!
//
//
// NOTE - even though the line
//
//	import att.research.yoix.*;
//
// works for this simple example, javac will sometimes get confused
// if your module source code is kept with the source code for the
// att.research.yoix package. If you run into strange javac errors
// move your source to a different directory and/or explicitly name
// the att.research.yoix classes that your module needs to import.
// For example, changing
//
//	import att.research.yoix.*;
//
// to
//
//	import att.research.yoix.YoixModule;
//	import att.research.yoix.YoixConstants;
//	import att.research.yoix.YoixObject;
//
// would be sufficient for our simple example.
//

public abstract
class ExampleModule extends YoixModule

    implements YoixConstants

{

    public static String  $MODULENAME = "ExampleModule";
    public static String  $MODULEVERSION = "2.0";

    public static Object  $module[] = {
    //
    // NAME                  ARG        COMMAND     MODE   REFERENCE
    // ----                  ---        -------     ----   ---------
       $MODULENAME,          "3",       $LIST,      $RORO, $MODULENAME,
       "DUMMY_CONSTANT",     "12",      $INTEGER,   $LR__, null,
       "property",           "1",       $BUILTIN,   $LR_X, null,
       "version",            "0",       $BUILTIN,   $LR_X, null,
    };

    ///////////////////////////////////
    //
    // ExampleModule Methods
    //
    ///////////////////////////////////

    public static YoixObject
    property(YoixObject arg[]) {

	String  value = null;

	if (arg[0].isString())
	    value = System.getProperty(arg[0].stringValue());
	else VM.badArgument(0);

	return(YoixObject.newString(value));
    }


    public static YoixObject
    version(YoixObject arg[]) {

	String  value = null;

	if (arg.length == 0)
	    value = System.getProperty("java.version");
	else VM.badCall();

	return(YoixObject.newString(value));
    }
}

