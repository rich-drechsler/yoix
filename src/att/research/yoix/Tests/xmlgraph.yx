import yoix.graph.*;
import yoix.stdio.*;

graph_str = "\
<graph name=g directed=1 strict=1 color=white label=Example>\n\
    <node_attributes color=blue />\n\
    <node name=node_0000 color=red />\n\
    <node name=node_0001 />\n\
    <node name=node_0002 color=green />\n\
    <node name=node_0003 />\n\
    <node name=node_0004 color=orange />\n\
    <edge name=edge_0003_0000 tail=node_0003 head=node_0000 color=yellow />\n\
    <edge name=edge_0000_0002 tail=node_0000 head=node_0002 color=green />\n\
    <edge name=edge_0002_0000 tail=node_0002 head=node_0000 color=red />\n\
    <edge name=edge_0001_0002 tail=node_0001 head=node_0002 color=blue />\n\
    <edge name=edge_0004_0000 tail=node_0004 head=node_0000 color=red />\n\
    <edge name=edge_0001_0004 tail=node_0001 head=node_0004 color=blue />\n\
</graph>\n\
";

g = xmlGraph(graph_str);

stdout.nextline = g.text();

stopColor(String col) {
    String color = attribute("color");
    printf("%s is %s\n", name, color);
    if(color === col) {
	printf("Found %s... stopping.\n", col);
	return(true);
    }
}

e = g.element("edge_0001_0004");
printf("First try 'grey':\n");
e.bfs(stopColor, "grey");
printf("Second try 'red':\n");
e.bfs(stopColor, "red");
