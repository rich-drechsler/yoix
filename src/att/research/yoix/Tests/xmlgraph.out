<graph name=g directed=1 strict=1 color=white label=example>
    <node_attributes color=blue />
    <node name=node_0000 color=red />
    <node name=node_0001 />
    <node name=node_0002 color=green />
    <node name=node_0003 />
    <node name=node_0004 color=orange />
    <edge name=edge_0003_0000 tail=node_0003 head=node_0000 color=yellow />
    <edge name=edge_0000_0002 tail=node_0000 head=node_0002 color=green />
    <edge name=edge_0002_0000 tail=node_0002 head=node_0000 color=red />
    <edge name=edge_0001_0002 tail=node_0001 head=node_0002 color=blue />
    <edge name=edge_0004_0000 tail=node_0004 head=node_0000 color=red />
    <edge name=edge_0001_0004 tail=node_0001 head=node_0004 color=blue />
</graph>

First try 'grey':
edge_0001_0004 is blue
edge_0004_0000 is red
edge_0000_0002 is green
edge_0002_0000 is red
Second try 'red':
edge_0001_0004 is blue
edge_0004_0000 is red
Found red... stopping.
